consensus/consortium/v2/consortium.go

```diff
@@ -899,18 +980,47 @@ func (c *Consortium) Finalize(chain consensus.ChainHeaderReader, header *types.H
 			return err
 		}
 
-		if len(checkpointValidator) != len(extraData.CheckpointValidators) {
-			return errMismatchingEpochValidators
-		}
-
-		for i := range checkpointValidator {
-			if checkpointValidator[i].Address != extraData.CheckpointValidators[i].Address {
-				return errMismatchingEpochValidators
+		// If isTripp and new period, read all validator candidates and
+		// their amounts, check with stored data in header
+		if c.IsTrippEffective(chain, header) {
+			if len(blockProducers) != len(extraData.BlockProducers) {
+				return errMismatchingValidators
+			}
+			for i := range blockProducers {
+				if blockProducers[i] != extraData.BlockProducers[i] {
+					return errMismatchingValidators
+				}
+			}
+			if isPeriodBlock {
+				if len(checkpointValidators) != len(extraData.CheckpointValidators) {
+					return errMismatchingValidators
+				}
+				for i := range checkpointValidators {
+					if checkpointValidators[i].Address != extraData.CheckpointValidators[i].Address {
+						return errMismatchingValidators
+					}
+					if !checkpointValidators[i].BlsPublicKey.Equals(extraData.CheckpointValidators[i].BlsPublicKey) {
+						return errMismatchingValidators
+					}
+					if checkpointValidators[i].Weight != extraData.CheckpointValidators[i].Weight {
+						return errMismatchingValidators
+					}
+				}
+			}
+		} else {
+			if len(checkpointValidators) != len(extraData.CheckpointValidators) {
+				return errMismatchingValidators
 			}
 
-			if isShillin {
-				if !checkpointValidator[i].BlsPublicKey.Equals(extraData.CheckpointValidators[i].BlsPublicKey) {
-					return errMismatchingEpochValidators
+			for i := range checkpointValidators {
+				if checkpointValidators[i].Address != extraData.CheckpointValidators[i].Address {
+					return errMismatchingValidators
+				}
+
+				if c.chainConfig.IsShillin(header.Number) {
```

Comment:
As this does not change during iteration, I think it's better to move this outside of the loop.